datafusion/functions-array/src/lib.rs

```diff
@@ -98,8 +103,190 @@ pub mod expr_fn {
+            Box::new(empty::array_empty_udf) as _,
+        );
+        functions.insert(
+            String::from("array_length"),
+            Box::new(length::array_length_udf) as _,
+        );
+        functions.insert(String::from("flatten"), Box::new(flatten::flatten_udf) as _);
+        functions.insert(
+            String::from("array_sort"),
+            Box::new(sort::array_sort_udf) as _,
+        );
+        functions.insert(
+            String::from("array_repeat"),
+            Box::new(repeat::array_repeat_udf) as _,
+        );
+        functions.insert(
+            String::from("array_resize"),
+            Box::new(resize::array_resize_udf) as _,
+        );
+        functions.insert(
+            String::from("array_reverse"),
+            Box::new(reverse::array_reverse_udf) as _,
+        );
+        functions.insert(
+            String::from("array_distinct"),
+            Box::new(set_ops::array_distinct_udf) as _,
+        );
+        functions.insert(
+            String::from("array_intersect"),
+            Box::new(set_ops::array_intersect_udf) as _,
+        );
+        functions.insert(
+            String::from("array_union"),
+            Box::new(set_ops::array_union_udf) as _,
+        );
+        functions.insert(
+            String::from("array_position"),
+            Box::new(position::array_position_udf) as _,
+        );
+        functions.insert(
+            String::from("array_positions"),
+            Box::new(position::array_positions_udf) as _,
+        );
+        functions.insert(
+            String::from("array_remove"),
+            Box::new(remove::array_remove_udf) as _,
+        );
+        functions.insert(
+            String::from("array_remove_all"),
+            Box::new(remove::array_remove_all_udf) as _,
+        );
+        functions.insert(
+            String::from("array_remove_n"),
+            Box::new(remove::array_remove_n_udf) as _,
+        );
+        functions.insert(
+            String::from("array_replace"),
+            Box::new(replace::array_replace_udf) as _,
+        );
+        functions.insert(
+            String::from("array_replace_all"),
+            Box::new(replace::array_replace_all_udf) as _,
+        );
+        functions.insert(
+            String::from("array_replace_n"),
+            Box::new(replace::array_replace_n_udf) as _,
+        );
+
+        // TODO: Add more builtin functions here
+        Mutex::new(functions)
+    })
+}
+
+// Get an UDF by name
+//
+// Replace with `get_udf`
+// fn get_function_meta(&self, name: &str) -> Option<Arc<ScalarUDF>> {
+//     self.state.scalar_functions().get(name).cloned()
+// }
+pub fn get_array_udf(name: &str) -> Option<Arc<ScalarUDF>> {
+    array_functions().lock().unwrap().get(name).map(|f| f())
+}
+
+/// Register a single new UDF, so the user can register their own functions
+///
+/// Repalce old regsiter_udf
+pub fn register_array_udf(name: &str, udf: ScalarFactory) -> Option<ScalarFactory> {
```

Comment:
I really don't like that this is some global mutable state that can be changed by the user. I think this will be very confusing as it both may have unintended side effects as well as not being clear if users should register their functions with the SessionState or this global map

I think a much simpler to reason about interface would be that the list of functions in `datafusion/functions` is immutable. For example something like 

```rust
pub fn array_functions() -> &'static <HashMap<String, ScalarFactory> {
```

Reply:
> both may have unintended side effects as well as not being clear if users should register their functions with the SessionState or this global map

My intention is to replace the old one, so there is only a single map that the user need to care about.